core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java

```diff
@@ -272,13 +309,33 @@ public void runClusterServices(@Nonnull Stack stack, @Nonnull Cluster cluster, M
         }
     }
 
+    private void addPrimaryGatewayToCandidatesIfNeeded(InstanceMetaData primaryGwImd, Set<Node> reachableCandidates) {
+        boolean candidatesMissingPrimaryGateway = reachableCandidates.stream().noneMatch(node ->
+                StringUtils.equals(node.getHostname(), primaryGwImd.getDiscoveryFQDN()));
+        if (candidatesMissingPrimaryGateway) {
+            Node primaryGatewayNode = new Node(primaryGwImd.getPrivateIp(), primaryGwImd.getPublicIp(), primaryGwImd.getInstanceId(),
+                    primaryGwImd.getInstanceGroup().getTemplate().getInstanceType(), primaryGwImd.getDiscoveryFQDN(),
+                    primaryGwImd.getInstanceGroupName());
+            // in case of upscale we should add primary gateway to candidates
+            reachableCandidates.add(primaryGatewayNode);
```

Comment:
I also would log the fact here that the primary gateway needed to be added to the candidates.